### PR TITLE
[Foundation] Set 'sentRequest' when sending a request in NSUrlSessionHandler.

### PR DESCRIPTION
**Pull request by rolfbjarne**
Fixes this compiler warning:

> [..]/external/mono/mcs/class/System.Net.Http/HttpClientEx.cs(50,8): warning CS0649: Field `Foundation.NSUrlSessionHandler.sentRequest' is never assigned to, and will always have its default value `false'

However it changes the runtime behavior, and we'll now throw an exception in
cases that we accepted before:

* `sentRequest` is only read in `EnsureModifiability ()`, which throws an
  exception if `sentRequest` is true.
* Previously `sentRequest` was never set (thus the compiler warning), which
  meant `EnsureModifiability` would never throw an exception.
* Looking at the similar `CFNetworkHandler` (which has the identical field and
  methods), it seems that the intended behavior is to set `sentRequest` in
  `SendAsync`, and then `EnsureModifiability` is called whenever a property is
  set to ensure the property isn't set too late (and any change would be
  ignored because the request was already sent).
* This means that previously setting any property after the request was sent
  would not throw any exceptions (even though the change would be ignored),
  while with this change we'd start throwing exceptions.

### CODE REVIEW

**Comment by rolfbjarne** [comment]
An alternative is to remove the whole `sentRequest` / `EnsureModifability` logic, which would no change any behavior, and at the same time make the code less confusing (and fix the compiler warning).

**Comment by chamons** [comment]
I don't know enough about the networking stack, so I'll fall on whatever @spouliot says on this.

**Comment by mandel-macaque** [comment]
@spouliot hmm I cannot believe I missed this. Looking into what this really means regarding to the old handler. We don't want to add new exceptions to apps that did not have them.

Also @rolfbjarne we have two diff implementations of the CFNetworkHandler, the mac and the iOS version. So we should make sure all of the do the same.

**Comment by mandel-macaque** [comment]
Lets block this until I add some tests for it.

**Comment by mandel-macaque** [comment]
Done the tests an everything looks ok. We can add them in a diff PR.

**Comment by rolfbjarne** [comment]
You can add the tests to this PR by pushing to my branch (I've checked the "Allow edits from maintainer" checkbox).

**Comment by mandel-macaque** [comment]
Super, I'll push the tests there for you

File: tests/monotouch-test/HttpClient/HttpClientTest.cs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+				Assert.DoesNotThrow (() => handler.AllowAutoRedirect = !handler.AllowAutoRedirect);
+				Task.Factory.StartNew (() => {
+					var token = new CancellationTokenSource ();
+					client.SendAsync (request, token.Token).Wait (WaitTimeout);
+					Assert.Throws<InvalidOperationException> (() => handler.AllowAutoRedirect = !handler.AllowAutoRedirect);
+					// cancel to ensure that we do not have side effects
+					token.Cancel ();
+				});
+			}
+		}
```
Comment (spouliot) [inline tests/monotouch-test/HttpClient/HttpClientTest.cs]: as discussed (1:1) we should avoid code duplication by using extension methods to expose properties like `AllowAutoRedirect` that exists in many handlers (but not in the base class).

Comment (mandel-macaque) [inline tests/monotouch-test/HttpClient/HttpClientTest.cs]: Sorted.

File: src/Foundation/NSUrlSessionHandler.cs [hunk 2]
```diff
@@ -169,6 +169,8 @@ string GetHeaderSeparator (string name)
 #endif
 		protected override async Task<HttpResponseMessage> SendAsync (HttpRequestMessage request, CancellationToken cancellationToken)
 		{
+			Volatile.Write (ref sentRequest, true);
```
Comment (rolfbjarne) [inline src/Foundation/NSUrlSessionHandler.cs]: Why are you using a volatile write instead of just assigning to the variable?

Comment (mandel-macaque) [inline src/Foundation/NSUrlSessionHandler.cs]: I used Volatile following the patterns used in the other handlers to keep everything as close as possible.

**Pull request merged**